### PR TITLE
Allow LINQ Expressions for Binding Expressions

### LINKED ISSUE #1652: Can not bind to attached property from xaml
Repro:
```XAML
<CheckBox IsChecked="{Binding Path=(id:DragBehavior.IsEnabled), Mode=TwoWay, RelativeSource={RelativeSource Self}}"/>
```
```XAML
<CheckBox IsChecked="{Binding $self.(id:DragBehavior.IsEnabled)}"/>
```
```C#
public static readonly AvaloniaProperty IsEnabledProperty =
    AvaloniaProperty.RegisterAttached<Control, bool>("IsEnabled", typeof(DropBehavior), true, true, BindingMode.TwoWay);

public static bool GetIsEnabled(Control control)
{
    return (bool)control.GetValue(IsEnabledProperty);
}

public static void SetIsEnabled(Control control, bool value)
{
    control.SetValue(IsEnabledProperty, value);
}
```
Error:
```
Invalid attached property name.
```

Stack trace:
```
   at Avalonia.Data.Core.Parsers.ExpressionParser.ParseAttachedProperty(Reader r, List`1 nodes) in C:\projects\Avalonia\src\Avalonia.Base\Data\Core\Parsers\ExpressionParser.cs:line 137
   at Avalonia.Data.Core.Parsers.ExpressionParser.Parse(Reader r) in C:\projects\Avalonia\src\Avalonia.Base\Data\Core\Parsers\ExpressionParser.cs:line 41
   at Avalonia.Data.Core.ExpressionNodeBuilder.Build(String expression, Boolean enableValidation) in C:\projects\Avalonia\src\Avalonia.Base\Data\Core\ExpressionNodeBuilder.cs:line 20
   at Avalonia.Data.Binding.CreateSourceObserver(Object source, String path, Boolean enableDataValidation) in C:\projects\Avalonia\src\Markup\Avalonia.Markup\Data\Binding.cs:line 269
   at Avalonia.Data.Binding.Initiate(IAvaloniaObject target, AvaloniaProperty targetProperty, Object anchor, Boolean enableDataValidation) in C:\projects\Avalonia\src\Markup\Avalonia.Markup\Data\Binding.cs:line 105
   at Avalonia.AvaloniaObjectExtensions.Bind(IAvaloniaObject target, AvaloniaProperty property, IBinding binding, Object anchor) in C:\projects\Avalonia\src\Avalonia.Base\AvaloniaObjectExtensions.cs:line 211
   at Avalonia.Markup.Data.DelayedBinding.BindingEntry.Apply(IStyledElement control) in C:\projects\Avalonia\src\Markup\Avalonia.Markup\Markup\Data\DelayedBinding.cs:line 132
   at Avalonia.Markup.Data.DelayedBinding.ApplyBindings(IStyledElement control) in C:\projects\Avalonia\src\Markup\Avalonia.Markup\Markup\Data\DelayedBinding.cs:line 97
   at Avalonia.Markup.Data.DelayedBinding.ApplyBindings(Object sender, EventArgs e) in C:\projects\Avalonia\src\Markup\Avalonia.Markup\Markup\Data\DelayedBinding.cs:line 110
   at Avalonia.Markup.Xaml.PortableXaml.AvaloniaXamlObjectWriter.DelayedValuesHelper.ApplyAll() in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\PortableXaml\AvaloniaXamlObjectWriter.cs:line 198
   at Avalonia.Markup.Xaml.PortableXaml.AvaloniaXamlObjectWriter.OnAfterProperties(Object value) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\PortableXaml\AvaloniaXamlObjectWriter.cs:line 85
   at Portable.Xaml.XamlObjectWriterInternal.OnWriteEndObject() in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\PortableXaml\portable.xaml.github\src\Portable.Xaml\Portable.Xaml\XamlObjectWriter.cs:line 398
   at Portable.Xaml.XamlWriterInternalBase.WriteEndObject() in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\PortableXaml\portable.xaml.github\src\Portable.Xaml\Portable.Xaml\XamlWriterInternalBase.cs:line 260
   at Portable.Xaml.XamlServices.Transform(XamlReader xamlReader, XamlWriter xamlWriter, Boolean closeWriter) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\PortableXaml\portable.xaml.github\src\Portable.Xaml\Portable.Xaml\XamlServices.cs:line 128
   at Avalonia.Markup.Xaml.AvaloniaXamlLoader.LoadFromReader(XamlReader reader, AvaloniaXamlContext context, IAmbientProvider parentAmbientProvider) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\AvaloniaXamlLoader.cs:line 208
   at Avalonia.Markup.Xaml.AvaloniaXamlLoader.Load(Stream stream, Assembly localAssembly, Object rootInstance, Uri uri) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\AvaloniaXamlLoader.cs:line 184
   at Avalonia.Markup.Xaml.AvaloniaXamlLoader.Load(Type type, Object rootInstance) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\AvaloniaXamlLoader.cs:line 89
   at Avalonia.Markup.Xaml.AvaloniaXamlLoader.Load(Object obj) in C:\projects\Avalonia\src\Markup\Avalonia.Markup.Xaml\AvaloniaXamlLoader.cs:line 57
   at AvaloniaDemo.MainWindow.InitializeComponent() in C:\DOWNLOADS\GitHub\Dock\samples\AvaloniaDemo\MainWindow.xaml.cs:line 134
   at AvaloniaDemo.MainWindow..ctor() in C:\DOWNLOADS\GitHub\Dock\samples\AvaloniaDemo\MainWindow.xaml.cs:line 20
```

### PR DESCRIPTION
**Pull request by jkoritzinsky**
- What does the pull request do?
   - Add new API to allow users to create `ExpressionObserver`s with LINQ Expressions instead of strings.
   - Move string-based binding paths up to Avalonia.Markup.
   - Fix attached property resolution to always resolve the correct property.
- What is the current behavior?
   - `ExpressionObserver`s can only have their binding paths specified as strings.
   - The `AvaloniaPropertyAccessorPlugin` tries to resolve an attached property name without the XAML context, so it just makes a guess of which property is the correct property.
- What is the updated/expected behavior with this PR?
   - `ExpressionObserver`s can be created from LINQ Expressions, `ExpressionNode`s, or strings.
   - Attached property bindings now use the `AvaloniaPropertyAccessorNode` class instead with an already-resolved `AvaloniaProperty` that is resolved via a callback (i.e. XAML type lookup) when built from a string.
- How was the solution implemented (if it's not obvious)?
  - Stream Bindings in LINQ Expressions are pattern based. A method that is either an instance method or extension method without parameters named `StreamBinding` generates a stream binding node at that point in the expression.

Checklist:

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

If the pull request fixes issue(s) list them like this:

Fixes #1652.

### CODE REVIEW

**Review by grokys — CHANGES_REQUESTED** [review]
Hey @jkoritzinsky I really love this! It's something I intended to do from the beginning but never got round to.

Just a few things:

1. Indexers don't seem to be working. We should really be testing expressions in the unit tests to catch stuff like this
2. Should we also allow `Binding` to be created with expressions? `Binding` provides a few extra features like binding to the `DataContext` or to other controls that could be really useful along with strong typing. This could be a separate PR though.

**Comment by grokys** [comment]
Also, does this PR depend on #1668? A lot of the changes seem to be similar.

**Comment by jkoritzinsky** [comment]
#1668 is dependent on some of the changes in this PR. After this gets merged in the changes over there will clean up a lot.

File: tests/Avalonia.LeakTests/ExpressionObserverTests.cs
Comment (jkoritzinsky) [reply tests/Avalonia.LeakTests/ExpressionObserverTests.cs]: Leak Tests aren't running on CI currently. That's fixed as part of my Cake PR.

File: src/Markup/Avalonia.Markup/Markup/Parsers/ExpressionObserverBuilder.cs
Comment (jkoritzinsky) [reply src/Markup/Avalonia.Markup/Markup/Parsers/ExpressionObserverBuilder.cs]: Currently it's public so XAML layer classes (`MemberSelector` and `TreeDataTemplate`) can use it just like they were using `ExpressionObserver` beforehand.

**Comment by grokys** [comment]
This is going to conflict with #1694, which do you think it makes more sense to get merged first?

**Comment by jkoritzinsky** [comment]
I'm on vacation for a week, so probably #1694. Alternatively, all that's left to do here is the XML doc comments, so if you want to put those in we could merge this on approval.

**Comment by jkoritzinsky** [comment]
@grokys I've made all the requested changes. Can you take another review pass when possible?

File: src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs
Comment (grokys) [inline src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs]: This applies to quite a few places, but elsewhere we use `_field` with an underscore for fields to follow the .net core coding guidelines.

File: tests/Avalonia.Base.UnitTests/Data/Core/ExpressionObserverTests_ExpressionTree.cs
Comment (grokys) [inline tests/Avalonia.Base.UnitTests/Data/Core/ExpressionObserverTests_ExpressionTree.cs]: Oh nice! I was just about to check if this works!

File: src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs
Comment (grokys) [inline src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs]: Does this need to be a separate class? Could this and `ExpressionVisitorNodeBuilder` be combined into a single class?

File: src/Avalonia.Base/Data/Core/ExpressionObserver.cs
Comment (grokys) [inline src/Avalonia.Base/Data/Core/ExpressionObserver.cs]: Usually, methods come after ctors. I understand why you've done it like this (every ctor has an accompanying static creation method) but I'd rather stick to our standard ordering.

**Review by grokys — CHANGES_REQUESTED** [review]
Mainly a few nits here, but I do wonder about my point 2 above: by making `ExpressionObserver` only handle expressions and pushing string expressions into `Binding`, we're making `ExpressionNode` part of the public API when it feels like an implementation detail.

What would be the cons of making `ExpressionObserver` handle strings as well?

File: src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs
Comment (jkoritzinsky) [reply src/Avalonia.Base/Data/Core/Parsers/ExpressionTreeParser.cs]: I could combine it, but I like having the visitor implementation in its own class. Keeps the messy logic of the visitor separated from the simpler logic in `ExpressionTreeParser`.

**Comment by jkoritzinsky** [comment]
I have a few reasons why I like having the string support outside of `ExpressionObserver`.

1. By having the string support in the Markup layer, the `Reader` class moves up as well (enabling #1668)
2. If we have string support in the base layer as a constructor (as done currently), then we make string bindings the easiest path of use. I'd like to push advanced users that want to directly use `ExpressionObserver` to use type-safe binding expressions if possible.
3. By making `ExpressionNode` and subclasses part of the API, we allow advanced users who want to customize the binding system (which is already supported on `ExpressionObserver` in master) to create custom binding nodes if they so wish.
4. Everything in `Avalonia.Data.Core` is already an implementation detail namespace. I don't think exposing anything in that namespace publicly is necessarily bad.

File: src/Avalonia.Base/Data/Core/ExpressionObserver.cs
Comment (jkoritzinsky) [reply src/Avalonia.Base/Data/Core/ExpressionObserver.cs]: I've moved them to after the constructors.

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Ok, yes those sound like good arguments for doing it this way 👍

**Pull request merged**
